app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -0,0 +1,248 @@
+      ethErrors.rpc.invalidParams({
+        message: `Expected single, object parameter. Received:\n${req.params}`,
+      }),
+    );
+  }
+
+  const { origin } = req;
+
+  const {
+    chainId,
+    chainName = null,
+    blockExplorerUrls = null,
+    nativeCurrency = null,
+    rpcUrls,
+  } = req.params[0];
+
+  const otherKeys = Object.keys(
+    omit(req.params[0], [
+      'chainId',
+      'chainName',
+      'blockExplorerUrls',
+      'iconUrls',
+      'rpcUrls',
+      'nativeCurrency',
+    ]),
+  );
+
+  if (otherKeys.length > 0) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Received unexpected keys on object parameter. Unsupported keys:\n${otherKeys}`,
+      }),
+    );
+  }
+
+  const firstValidRPCUrl = Array.isArray(rpcUrls)
+    ? rpcUrls.find((rpcUrl) => validUrl.isHttpsUri(rpcUrl))
+    : null;
+
+  const firstValidBlockExplorerUrl =
+    blockExplorerUrls !== null && Array.isArray(blockExplorerUrls)
+      ? blockExplorerUrls.find((blockExplorerUrl) =>
+          validUrl.isHttpsUri(blockExplorerUrl),
+        )
+      : null;
+
+  if (!firstValidRPCUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
+      }),
+    );
+  }
+
+  if (blockExplorerUrls !== null && !firstValidBlockExplorerUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected null or array with at least one valid string HTTPS URL 'blockExplorerUrl'. Received: ${blockExplorerUrls}`,
+      }),
+    );
+  }
+
+  const _chainId = typeof chainId === 'string' && chainId.toLowerCase();
+
+  if (!isPrefixedFormattedHexString(_chainId)) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected 0x-prefixed, unpadded, non-zero hexadecimal string 'chainId'. Received:\n${chainId}`,
+      }),
+    );
+  }
+
+  if (!isSafeChainId(parseInt(_chainId, 16))) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Invalid chain ID "${_chainId}": numerical value greater than max safe value. Received:\n${chainId}`,
+      }),
+    );
+  }
+
+  if (CHAIN_ID_TO_NETWORK_ID_MAP[_chainId]) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `May not specify default MetaMask chain.`,
+      }),
+    );
+  }
+
+  const existingNetwork = findCustomRpcBy({ chainId: _chainId });
+
+  if (existingNetwork !== null) {
+    const currentNetwork = getCurrentNetwork();
+    if (currentNetwork.chainId === _chainId) {
+      res.result = null;
+      return end();
+    }
+    try {
+      await updateRpcTarget(
+        await requestUserApproval({
+          origin,
+          type: MESSAGE_TYPE.SWITCH_ETHEREUM_CHAIN,
+          requestData: {
+            rpcUrl: existingNetwork.rpcUrl,
+            chainId: existingNetwork.chainId,
+            nickname: existingNetwork.nickname,
+            ticker: existingNetwork.ticker,
+          },
```

Comment:
Are we unable to support the updating of the block-explorer url?

Reply:
when we get here the user will have already added the chain to their MetaMask, and the block explorer URL will have been set. This requestData is set because updateRpcTarget requires it.